### ISSUE TITLE
test_bgl_unfair fails consistently on Windows with Python3.5

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#265]
Example: https://dev.azure.com/responsibleai/fairlearn/_build/results?buildId=3657&view=logs&jobId=a846d25a-e32c-5640-1b53-e815fab94407&j=f2ae15cd-30d2-5fb2-07f7-558e9be6b891&s=778b61a2-25e7-5600-2cb0-dd2dc8e23627

The affected code:
```
assert logging_all_close([[3.2, 11.2],
                              [-3.47346939, 10.64897959],
                              [-2.68, 10.12],
                              [-1.91764706, 9.61176471],
                              [-1.18461538,  9.12307692],
                              [-0.47924528,  8.65283019],
                              [0.2, 0.7]],
                             all_predict)
```
and only the first entry fails.
```
E        +  where False = logging_all_close([[3.2, 11.2], [-3.47346939, 10.64897959], [-2.68, 10.12], [-1.91764706, 9.61176471], [-1.18461538, 9.12307692], [-0.47924528, 8.65283019], ...], [array([ 3.03010885, 11.2       ]), array([-3.47346939, 10.64897959]), array([-2.68, 10.12]), array([-1.91764706,  9.61176471]), array([-1.18461538,  9.12307692]), array([-0.47924528,  8.65283019]), ...])

test\unit\reductions\grid_search\test_grid_search_regression.py:72: AssertionError
---------------------------- Captured stdout call -----------------------------
a mismatches:  [3.2]
b mismatches:  [3.03010885]
mismatch indices:  (array([0], dtype=int64), array([0], dtype=int64))
```

**Comment by romanlutz** [comment]
According to Miro this may be because it's at the edge of the grid where one of the classes is ignored, and that may lead to imbalance resulting in multiple solutions with the same least squares loss. So that means it's actually undetermined. I'm putting in some tolerance to accept both to unblock myself, but we'll need to see whether the hypothesis can be verified.

**Comment by riedgar-ms** [comment]
That looks pretty similar to the transient a week or so ago. Same item, same rough tolerance required. Which is rather large (on the order of 10%). Even if this is a case of two equally good solutions, it seems odd that we've only seen it recently. AFAICT, the `LinearRegression` solver is supposed to be fully deterministic (it wraps SciPy, which in turn wraps LAPACK), so even if there are two solutions which are equally good it should still always chose the same one of them.

**Comment by MiroDudik** [comment]
Don't increase tolerance. The tolerance should still be at essentially the
numerical level.

If you want to unblock then modify the test so that it ignores the two
extremes (where one group gets the weight zero and the other group weight
one) and only do comparison of the inner elements of the array.

**Comment by riedgar-ms** [comment]
@romanlutz has already modified the test to make the first element a special case. If the last element starts giving us trouble, we can do the same for the last. This issue is mentioned in the explanatory comment.

**Comment by riedgar-ms** [comment]
This still leaves the question about what causes the selected solution to change.... however, if there are multiple equally good solutions, then I suppose the answer is 'literally anything in the numeric stack' which would be a wild goose chase (since potential important differences would include the exact patches on each VM).

Perhaps the real issue is a need for more sophisticated testing, rather than what I created here - a golden value test which just makes sure that the algorithm keeps giving the same answers.

**Comment by MiroDudik** [comment]
Agreed about improving tests. I don't think that a comparison to the golden value is a bad idea, we should just make sure that our estimator(s) (on top of which we implement the reductions) have a unique minimum. This is easily achieved by adding a bit of regularization.

So my suggestion for a systematic solution:
* use estimators from `test/unit/reductions/exponentiated_gradient/simple_learners.py`
* modify those estimators, so they can have a bit of regularization

I prefer this to using estimators from sklearn.linear_model--these allow regularization, so we could solve the uniqueness problem there too--but I'd prefer to avoid sklearn dependence in tests like these (we could even try to avoid numpy, but I'm not that paranoid).

**Comment by romanlutz** [comment]
Just to add to this: the "fix windows notebook" change that I just merged does not solve the overall problem, it just adds the special case handling back. Somehow it worked for a little while without it (probably because the affected platforms aren't part of the PR gate). The overall resolution remains to be carried out.

**Comment by riedgar-ms** [comment]
Isn't this fixed now @romanlutz ?

**Comment by riedgar-ms** [comment]
Since we're in the process of dropping Python 3.5, should we close this as a 'won't fix' ?

**Comment by romanlutz** [comment]
To my knowledge this occurs in py3.6 as well

**Comment by riedgar-ms** [comment]
Do you have a repro? Our nightly builds are both green right now.

**Comment by romanlutz** [comment]
Of course they are, because we put a special case in the test :-)

```
# TODO: investigate where the different outcomes for the first grid point are from, likely
    # due to some ignored data points at the edge resulting in another solution with the same
    # least squares loss (i.e. both solutions acceptable).
    # Reflects https://github.com/fairlearn/fairlearn/issues/265
    assert logging_all_close([[3.2, 11.2]], [all_predict[0]]) or \
        logging_all_close([[3.03010885, 11.2]], [all_predict[0]])

    assert logging_all_close([[-3.47346939, 10.64897959],
                              [-2.68, 10.12],
                              [-1.91764706, 9.61176471],
                              [-1.18461538,  9.12307692],
                              [-0.47924528,  8.65283019],
                              [0.2, 0.7]],
                             all_predict[1:])
```

**Comment by romanlutz** [comment]
https://github.com/fairlearn/fairlearn/blob/b07209058322d36e0962e25f82cd720ada650cde/test/unit/reductions/grid_search/test_grid_search_regression.py#L82

**Comment by MiroDudik** [comment]
As a reminder, although Python 3.6 is deprecated, the ill-defined optimization solution is still there, so I suggest we leave this issue open.

**Comment by romanlutz** [comment]
Plus it happens in 3.7 as I found out recently!

**Comment by romanlutz** [comment]
Very much still happening in 3.9 as I just reproduced. As @MiroDudik pointed out above this is due to an "ill-defined optimization solution."

**Comment by romanlutz** [comment]
Update: it appears to go away if I uprade to `numpy>2`. Are we ready to enforce such a thing?

**Comment by TamaraAtanasoska** [comment]
@romanlutz unfortunately not, it's just been a year since the release at least that seems to be the consensus recently discussed here #1491 . It would also help with other compatibilities and simplifying things, but sometime in the future
